designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,295 @@
+
+### Inference algorithm outputs
+
+The outputs from the inference algorithm necessarily differ according to the method.
+The inference algorithm either sends information to the callback directly, or, in the case
+of the HMC samplers, helper functions make use of  utility class `stan::services::utils::mcmc_writer`
+which assembles the outputs from the Stan model and the inference algorithm into
+a single row's worth of information sent to the writer callbacks.
+
+
+### Interface-level outputs
+
+The CmdStan interface dumps all configuration information into the Stan CSV file as a series of initial comment lines.
+The CmdStan argument parser uses indentation to indicate nested argument structure.
+Following the argument parser are several more comment lines of key-value pairs of information
+about the software version and compiler options used to compile the Stan model executable.
+To save a record of the inference run, we wish to store this in a structured format
+allowing for lists, maps, and nested lists and maps.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We propose to break down the information from the Stan inference algorithms into a series of output streams
+where each stream handles one kind of information and whose format is appropriate for further processing.
+Tabular data will be output either in CSV format (human readable) or using the Apache Arrow binary formats.
+For hierarchical data we will use JSON format.
+
+Tabular information includes:
+
+- The output from the Stan model class `write_array` function: the computed values for all Stan model variables on the constrained scale.
+- The Stan model parameter variables on the unconstrained scale returned by the `log_prob` function.
+- The inference algorithm state at each step of the iterative algorithm.
+
+Structured information includes:
+
+- The HMC tuning parameters:  metric and stepsize.  The HMC metric is a symmetric positive definite matrix, either a diagonal or a dense matrix.
+The former is reported as just the vector that is matrix diagonal, the latter is reported as the full matrix.  The HMC stepsize is a scalar.
+
+- Stan model variables schema: a list of all model variables including: block, name, type, shape, size, output column name(s)
+
+- Information about the inference run:  algorithm configuration, Stan version, model compiler version and compile options, and timing information.
+
+## Implementation
+
+To improve and generalize the outputs and output formats, we propose to:
+
+- implement a new set of algorithm-specific output handlers which
+will be passed in to the `stan::services` layer wrapper function and which
+will manage the necessary callback writer objects.
+
+- decouple the output format from the writer through the use of formatter callbacks.
```

Comment:
This gets at my earlier question a little, but I think it needs a bit more. Some of this may not be clear until implementation, but I would love a sketch of the callback API, and a bit on how the same information would be formatted differently if using the csv/legacy version and the newer arrow/json version

Reply:
let's talk.